dcap/ql/src/quote_enclave.rs

```diff
@@ -8,17 +8,167 @@
 //! has a mix up.  It uses the *verification* description for `sgx_ql_set_path`
 //! and the "generation" description for `sgx_qv_set_path`
 
+use crate::Error;
+use core::result::Result as CoreResult;
 use mc_sgx_core_sys_types::sgx_target_info_t;
 use mc_sgx_core_types::TargetInfo;
 use mc_sgx_dcap_ql_types::PathKind;
-use mc_sgx_dcap_types::{Quote3Error, RequestPolicy};
+use mc_sgx_dcap_types::RequestPolicy;
 use mc_sgx_util::ResultInto;
+use once_cell::sync::OnceCell;
 use std::{ffi::CString, os::unix::ffi::OsStrExt, path::Path};
 
+/// A convenience type alias for a `Result` which contains an [`Error`].
+pub type Result<T> = CoreResult<T, Error>;
+
+/// Initialization of the paths for the quoting enclaves and quote provider
+/// library
+///
+/// This can only be called once during process start up utilizing
+/// [`PathInitializer::with_paths()`] or [`PathInitializer::try_default()`].
+/// If a consumer of this crate does not explicitly initialize the paths, then
+/// they will be defaulted on the first call to an SGX function that needs the
+/// paths set.
+#[derive(Debug)]
+pub struct PathInitializer;
+static PATH_INITIALIZER: OnceCell<PathInitializer> = OnceCell::new();
```

Comment:
One idea that come up in talking with @remoun is we could pass the paths as a plain old data type to the functions that need them.  And set them in those function prior to calling the other sgx methods.  

This wouldn't account for threading, but would avoid the global singleton.

Reply:
offline discussion seemed to want to stick with a singleton style